src/core/ext/filters/client_channel/lb_policy/grpclb/grpclb.cc

```diff
@@ -1163,36 +1163,56 @@ static int glb_pick_locked(grpc_exec_ctx *exec_ctx, grpc_lb_policy *pol,
                            "won't work without it. Failing"));
     return 0;
   }
-
   glb_lb_policy *glb_policy = (glb_lb_policy *)pol;
-  bool pick_done;
-
+  bool pick_done = false;
   if (glb_policy->rr_policy != NULL) {
-    if (GRPC_TRACER_ON(grpc_lb_glb_trace)) {
-      gpr_log(GPR_INFO, "grpclb %p about to PICK from RR %p",
-              (void *)glb_policy, (void *)glb_policy->rr_policy);
-    }
-    GRPC_LB_POLICY_REF(glb_policy->rr_policy, "glb_pick");
+    const grpc_connectivity_state rr_connectivity_state =
+        grpc_lb_policy_check_connectivity_locked(exec_ctx,
+                                                 glb_policy->rr_policy, NULL);
+    // The glb_policy->rr_policy may have transition to SHUTDOWN but the
+    // callback registered to capture this event
+    // (glb_rr_connectivity_changed_locked) may not have been invoked yet. We
+    // need to make sure we aren't trying to pick from a RR policy instance
+    // that's in shutdown.
+    if (rr_connectivity_state == GRPC_CHANNEL_SHUTDOWN) {
+      if (GRPC_TRACER_ON(grpc_lb_glb_trace)) {
+        gpr_log(GPR_INFO,
+                "grpclb %p NOT picking from from RR %p: RR conn state=%s",
+                (void *)glb_policy, (void *)glb_policy->rr_policy,
+                grpc_connectivity_state_name(rr_connectivity_state));
+      }
+      // Attempt to switch over to a pending update.
+      rr_handover_locked(exec_ctx, glb_policy);
+      add_pending_pick(&glb_policy->pending_picks, pick_args, target, context,
```

Comment:
If we are going to call `rr_handover_locked()` here, then let's add the pending pick *before* we do that, just to make sure that we don't somehow wind up processing the connectivity callback before we add the pending pick.  I don't think that can actually happen right now, but better safe than sorry -- the result would be that we'd leave an entry in pending picks that would never actually get processed.

Reply:
Done